src/python/pants/util/objects.py

```diff
@@ -52,6 +55,184 @@ def __getnewargs__(self):
   return DataType
 
 
+class TypedDatatypeClassConstructionError(Exception):
+
+  # TODO(cosmicexplorer): make some wrapper exception class to make this kind of
+  # prefixing easy (maybe using a class field format string?).
+  def __init__(self, type_name, msg, *args, **kwargs):
+    full_msg =  "while trying to generate typed dataype '{}': {}".format(
+      type_name, msg)
+    super(TypedDatatypeClassConstructionError, self).__init__(
+      full_msg, *args, **kwargs)
+
+
+class TypedDatatypeInstanceConstructionError(Exception):
+
+  def __init__(self, type_name, msg, *args, **kwargs):
+    full_msg = "in constructor of type '{}': {}".format(type_name, msg)
+    super(TypedDatatypeInstanceConstructionError, self).__init__(
+      full_msg, *args, **kwargs)
+
+
+class TypeCheckError(TypedDatatypeInstanceConstructionError):
+
+  def __init__(self, type_name, msg, *args, **kwargs):
+    formatted_msg = "type check error:\n{}".format(msg)
+    super(TypeCheckError, self).__init__(
+      type_name, formatted_msg, *args, **kwargs)
+
+
+class TypeDecl(AbstractClass):
+
+  class ConstructionError(Exception):
+    pass
+
+  @abstractmethod
+  def matches_value(self, val):
+    """Return whether or not the argument matches the type described by this
+  class.
+    """
+
+  @abstractmethod
+  def as_union(self):
+    """Return a version of this object which is an instance of Union."""
+
+  def compose(self, rhs):
+    """Return a TypeDecl which matches either this or another TypeDecl."""
+    self_types = self.as_union().types
+    rhs_types = rhs.as_union().types
+    all_types = self_types + rhs_types
+    return Union(*all_types)
+
+
+class SimpleTypeDecl(datatype('SimpleTypeDecl', ['matching_type']), TypeDecl):
+
+  def __new__(cls, matching_type):
+    if not isinstance(matching_type, type):
+      raise cls.ConstructionError(
+        "argument should be a type: '{}'".format(matching_type))
+
+    return super(SimpleTypeDecl, cls).__new__(cls, matching_type)
+
+  def matches_value(self, val):
+    return isinstance(val, self.matching_type)
+
+  def as_union(self):
+    return Union(self.matching_type)
+
+
+class Union(datatype('Union', ['types']), TypeDecl):
+
+  def __new__(cls, *types):
+    if len(types) == 0:
+      raise cls.ConstructionError("at least one type must be provided to Union")
+    if not all([isinstance(ty, type) for ty in types]):
+      raise cls.ConstructionError(
+        "all arguments to Union should be types: '{}'".format(types))
+
+    return super(Union, cls).__new__(cls, types)
+
+  def matches_value(self, val):
+    return any([isinstance(val, ty) for ty in self.types])
+
+  def as_union(self):
+    return self
+
+
+def typed_datatype(type_name, field_decls):
+  """A wrapper over namedtuple which accepts a dict of field names and types.
+
+  This can be used to very concisely define classes which have fields that are
+  type-checked at construction.
```

Comment:
Do we want exact type matching, or allow subtypes? Currently, this allows subtypes. That might be tricky for products in v2. I think v2's type system is primarily based on exact matches of types, either as a singular type or as a type within a union.

If this for datatypes outside v2 that is probably fine. Either way, I think the entry point here should call out how types are matched.

If you want, you could use the existing `TypeConstraint` classes that are used by the engine rules. https://github.com/pantsbuild/pants/blob/master/src/python/pants/engine/addressable.py#L31

Reply:
I think the right move is probably to do exact type matching for the reasons you describe. I had seen the TypeConstraint classes but hadn't used them -- that seems like the right way to tackle this.

Reply:
Also, for union types, exact type matching means you can use a set instead of a list to record the types that could match (and determine whether the type matches instantly), which in itself sounds like a fantastic argument for exact type checking.

Reply:
Looked it over, and found `TypeConstraint` to be a drop-in replacement for the `TypeMatcher` I had, so used that instead. Had to move `TypeConstraint` into `objects.py` to avoid a cyclic dependency, but it was easy enough to move and passes tests locally as of ae70e02. The only difference is that it doesn't convert its `self._types` into a `frozenset` -- this seems like a tiny change so I will probably add that.